### PR TITLE
    Issue 5013 - backport of try_lock to 1.3.10

### PR DESCRIPTION
**Pull request by jchapma**
Description: A change was previously made to connection
    management where the connection lock was replaced with a try_lock.
    This change resulted in a performance improvement in connection handling.

    This change was merged to 1.4.1.6 but not to the 1.3.10 branch, this
    issue looks at backporting this change to 1.3.10.

    relates: https://github.com/389ds/389-ds-base/issues/3550

    Fixes: https://github.com/389ds/389-ds-base/issues/5013

    Reviewed by: ?

### CODE REVIEW

**Review by Firstyear — APPROVED** [review]

File: ldap/servers/slapd/conntable.c [hunk 1]
```diff
@@ -118,40 +136,60 @@ Connection *
...
-            PR_Unlock(ct->table_mutex);
-            if (c->c_mutex == NULL || c->c_pdumutex == NULL) {
-                c->c_mutex = NULL;
-                c->c_pdumutex = NULL;
-                slapi_log_err(SLAPI_LOG_ERR, "connection_table_get_connection", "PR_NewLock failed\n");
-                exit(1);
-            }
-        }
+        
+	if(c->c_state == CONN_STATE_INIT) {
```
Comment (tbordaz) [inline ldap/servers/slapd/conntable.c]: I am not sure this block is valid.
c_mutex has been allocated during connection_table_new and c_state=CONN_STATE_INIT.

So we should realloc c_mutex only if CONN_STATE_FREE. Shouldn't the test be c->c_state==CONN_STATE_FREE ?

Also, once the connection table has been allocated, all slots should be CONN_STATE_INIT and c_mutex allocated. When is it necessary to reinit (alloc c_mutex and c_pdumutex) again ?

Comment (jchapma) [reply ldap/servers/slapd/conntable.c]: > 

@tbordaz 
Yes, you are correct in both counts. We dont need to realloc c_mutex at all, it was some legacy stuff that I forgot to remove...  Thanks

**Review by tbordaz — COMMENTED** [review]
@jchapma the code looks excellent, thanks. However I would like to have deeper review.
few questions

- Do you intend to push the fix upstream (branch 1.4.3, 1.4.4 ?)
- One of the expectation of the fix is to boost the performance. Did you do some perf measurement to confirm the gain ?

**Comment by jchapma** [comment]
> @jchapma the code looks excellent, thanks. However I would like to have deeper review. few questions

@tbordaz, it is excellent, all credit goes to the author @Firstyear, I am just backporting it.
> 
>     * Do you intend to push the fix upstream (branch 1.4.3, 1.4.4 ?)
From what I understand, this change was merged into 1.4.1, so its just the backport to 1.3.10 that we are evaluating.

>     * One of the expectation of the fix is to boost the performance. Did you do some perf measurement to confirm the gain ?
I done some performance testing using the 10k connection test, here are some comparisons of 1.3.10 and various backports:

**1.3.10 with no modifications**
Peak Concurrent Connections:   10000
Total Connections:             10000         (47.28/sec)  (2836.90/min)
 - LDAP Connections:           10000         (47.28/sec)  (2836.90/min)
 - LDAPI Connections:          0             (0.00/sec)  (0.00/min)
 - LDAPS Connections:          0             (0.00/sec)  (0.00/min)
Connections Reset By Peer:     0

**1.3.10 with accept_thread backport**
Peak Concurrent Connections:   10000
Total Connections:             10000         (86.49/sec)  (5189.67/min)
 - LDAP Connections:           10000         (86.49/sec)  (5189.67/min)
 - LDAPI Connections:          0             (0.00/sec)  (0.00/min)
 - LDAPS Connections:          0             (0.00/sec)  (0.00/min)
Connections Reset By Peer:     0

**1.3.10 with accept_thread and try_lock backports**
Peak Concurrent Connections:   10000
Total Connections:             10000         (137.89/sec)  (8273.27/min)
 - LDAP Connections:           10000         (137.89/sec)  (8273.27/min)
 - LDAPI Connections:          0             (0.00/sec)  (0.00/min)
 - LDAPS Connections:          0             (0.00/sec)  (0.00/min)
Connections Reset By Peer:     0

**Comment by tbordaz** [comment]
@jchapma, yes my mistake I was thinking to 1.3.10 rather than 1.4.3. So this patch is for a backport in 1.3.10 branch not a hotfix.

Regarding the performance, those metrics are promising as this patch is showing huge gain of 60% (137 vs 86).
As there is a pending issue with accept_thread, it would rather make sense to compare 1.4.3 vanilla vs 1.4.3+5013.

The gain is with a specific load (how fast can we establish new connections). This patch should also improve regular server activity as #op /sec. Could you also run some tests search rate 1.4.3 vs 1.4.3+5013 ?

**Comment by Firstyear** [comment]
There were approved, is there anything else to do?

**Comment by tbordaz** [comment]
@Firstyear, 1.4.3 introduces a set of fixes around connections handling (connection free list, transition to pthread mutex, dedicated accept thread, and likely others). Backporting pthread mutex looks a simple fix but we agreed that it is a waste of time to improve 1.3.10 while 1.4.3-4 and 2.0 are the active branch. If facing a connection related issue on 1.3.10, it is better to upgrade to 1.4.3 or 2.0 rather than taking partial improvements in 1.3.10.

In short, IMHO  this ticket should be close at wontfix.

**Comment by Firstyear** [comment]
Cool, that's fine by me. Closing and if we need it @jchapma can re-open it :)
